### PR TITLE
🐛 [RFR for API Sources] Fix bug where checkpoint reader stops syncing too early if first partition is complete

### PR DESCRIPTION
**Pull request by brianjlai**
Fixes https://github.com/airbytehq/airbyte-internal-issues/issues/8668

## What

Fixes a bug where if for a substream, the first parent has already successfully synced and had the success terminal value, then we would stop iterating over parent slices too soon and miss unfinished parent record slices.

Note: This does not have any customer impact because we've gated substreams off for all sources. I just noticed this in another PR.

## How

We need to add logic to skip over completed streams if the state is set to the sentinel value indicating a successful sync. And we continue reading the slice iterator until we find an incomplete parent slice or finish going over all parents in which case we are done syncing

I originally found this while implementing https://github.com/airbytehq/airbyte/pull/39450 , but i split out the fix to unblock lwo-code RFR which would have been affected by the bug

## Review guide

it's two files

## User Impact

None

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:115]: nit: 
```python
if not self._current_slice:
    ...
```

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:124]: There are many places where this part: `{"__ab_full_refresh_sync_complete": True}` of code could be reused, can we have the `constant` assigned to this value and then reused instead of duplicating the code?

like: 

```python
FULL_REFRESH_SYNC_COMPLETE: Dict = {"__ab_full_refresh_sync_complete": True}
```

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:139]: Can we have the dedicated `setter` and assign the value to the instance variable: `self._current_slice`, like:
```python
def set_current_slice(self, value) -> None:
   self._current_slice = value
```
Then reuse this within the code parts.

This clears up the code and makes it easier to follow.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:135]: This is preferable to do within the external method, if possible, since we don't return anything, it's perfectly aligned.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:124]: ah yes. that was the same comment on this [PR](https://github.com/airbytehq/airbyte/pull/39450) from Alex. I probably forked this PR out before I made that change to the constants, but will fix

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:115]: The difference here is subtle. But the intent here was that `StreamSlice` technically dervies from a mapping which could theoretically be an empty `{}` mapping.

To use a real example, let's say we have a basic stream that doesn't have any stream slicing and just returns a basic empty StreamSlice == {}.

The first iteration will work, but on the second page, we will enter this loop and move to the next slice which will return StopIteration. What we actually should do is move to the else and continue paginating.

So in this case `StreamSlice is None` is not equivalent to `not StreamSlice`

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:139]: yeah i can actually just transition this into python getter/setter property pattern and replace, not a problem

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:135]: i'm not sure extracting this actually makes it more readable. I experimented a bit with this:
```
    def _find_next_unfinished_slice(self) -> StreamSlice:
        next_candidate_slice = None
        has_more = True
        while has_more:
            next_candidate_slice = self._get_next_slice()
            state_for_slice = self._cursor.select_state(next_candidate_slice)
            has_more = state_for_slice == FULL_REFRESH_COMPLETE_STATE
        return next_candidate_slice
```

It does mostly work, but we do actually need to pass back both `state_for_slice` in addition to `next_candidate_slice`. We could return it as a tuple, but it just adds more layers to where the variables need to be passed around and used. For a relatively small block of code, i think leaving it as is is cleaner

**Review by girarda — APPROVED** [review]
approved pending @bazarnov's approval

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (bazarnov) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:142]: No problem, but if it's crashed at certain point it's easier to find the actual root cause, from the method name rather than get familiar with the pretty complex logic of the whole method. Not blocking on this!

**Review by bazarnov — APPROVED** [review]

**Comment by bazarnov** [comment]
LGTM! Thanks for the extra effort explaining the important moments on this subject! @brianjlai

**Pull request merged**
